#252/limit profile pictures

nataliiazab opened the pull request:
# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review on my code
- [x] I have made corresponding changes to the documentation (if any were needed)
- [x] My changes generate no new warnings or errors
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and previously existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

# Description

**Relates #252, #267 **  
This is to limit profile image to 1 per profile and delete all old ones stored in /project/default/storage/buckets/images
Additionally, the PR changed the CDN URL based on the environment (production vs. development) to also view images for local development.

### Files changed

1. **UploadImageInput.tsx**

- The code below fetches existing images for the user and deletes them if they exist. Then, it uploads the new image to supabase storage. This code ensures that there is only 1 profile picture per user, even if there was more than 1 picture stored on supabase. 
```
const { data: listData, error: listError } = await supabase.storage
            .from('images')
            .list(userId, { recursive: true });
```

```
 const oldImages = listData?.map((image) => image.name) || [];
          if (oldImages.length > 0) {
            const { error: deleteError } = await supabase.storage
              .from('images')
              .remove(oldImages.map((image) => userId + '/' + image));
```
```
  const { error: uploadError } = await supabase.storage
            .from('images')
            .upload(userId + '/' + imageName, file);
```

- The code dynamically assigns the base URL for accessing images based on the environment in which the application is running. In production, it uses main storage URL for the project, while in development, it uses a local server URL (currently it is hard-coded). This helps to fix the issue we've discussed in discord, that was also raised by Yurii - #267 

```
const CDN = process.env.NODE_ENV === 'production'
  ? 'https://undfcbmldjkujposixvn.supabase.co/storage/v1/object/public/images/'
  : 'http://127.0.0.1:54321/storage/v1/object/public/images/';
```

</summary>

2. **next.config.js**

- After changing CDN link, I've got a new error because next/image requires users to specify which domains are allowed for image sources in the next.config.js. By adding the local hostnames to the domains array in next.config.js, I was able to solve this.

```
images: {
  remotePatterns: [
    { protocol: 'https', hostname: 'undfcbmldjkujposixvn\\.supabase\\.co', pathname: '/storage/v1/object/public/images/**' },
    { protocol: 'http', hostname: '127.0.0.1', pathname: '/storage/v1/object/public/images/**' },
    { protocol: 'http', hostname: 'localhost', pathname: '/storage/v1/object/public/images/**' },
  ],
},
```



### UI changes

There are no changes to UI. However, I suggest creating a new ticket for profile image upload. When the new image is uploaded, there is no message for the user that it is successful and it is confusing if the upload worked at all (user can only reload the page to see the changed profile picture). I am happy to create one and work on it, if you think it is relevant :)


# Tests
No changes. Let me know if needed, I am still learning tests
updated on 8/5/2024: added userProfile.cy.js file to test if a user can edit their profile username and image

camelPhonso commented on components/form/UploadImageInput.tsx:
```diff
@@ -11,7 +11,9 @@ type UploadImageProps = {
-  'https://undfcbmldjkujposixvn.supabase.co/storage/v1/object/public/images/';
+  process.env.NODE_ENV === 'production'
+    ? 'https://undfcbmldjkujposixvn.supabase.co/storage/v1/object/public/images/'
+    : 'http://127.0.0.1:54321/storage/v1/object/public/images/';
```
This is genius! But I actually needed to make a couple of changes on my end to make this work:

In my `Studio URL` I needed to create a bucket by going onto `storage > new bucket`
![Screenshot 2024-07-29 at 22 21 44](https://github.com/user-attachments/assets/60f639d8-4144-48a8-94c3-65ce7529299b)

Before I did that the images were not showing anywhere, but after this step I was able to upload them successfully ✨ 

Did you need to do the same in your local environment?

camelPhonso replied:
That makes sense. We'll definitely need to record this somehow to make sure everyone has the same bucket once we merge these changes then.

Weirdly enough though I've just retried creating the bucket and uploading images and now they are showing up in the bucket but won't display on the page for some reason 🤔 they were showing just fine when I did my first review however.

nataliiazab replied:
Would you try again please? Mine are working ok at the moment

camelPhonso replied:
It was the `public` setting on my bucket that was giving trouble - sorry 😬

camelPhonso commented on components/form/UploadImageInput.tsx:
```diff
@@ -54,19 +56,44 @@ const UploadImageInput: React.FC<UploadImageProps> = ({
+            return;
+          }
+
+          const oldImages = listData?.map((image) => image.name) || [];
```
This all looks great but I'll share the same feedback I've given a few other contributors - let's lean on TypeScript and avoid using a fallback value (` || []`). 

If it TypeScript complains of the expected types it's preferable that we add an assertion wherever it makes sense to, but if the value coming out of any query is not what we expected we should let the code break to make it easier for us later if we need to debug.

camelPhonso commented on next.config.js:
```diff
@@ -5,11 +5,23 @@ const nextConfig = {
         pathname: '/storage/v1/object/public/images/**',
       },
+      //local development image host
+      {
+        protocol: 'http',
+        hostname: '127.0.0.1',
+        pathname: '/storage/v1/object/public/images/**',
+      },
+      {
+        protocol: 'http',
+        hostname: 'localhost',
+        pathname: '/storage/v1/object/public/images/**',
+      },
```
I'm actually a little lost on this one. The pathname doesn't really match my local bucket but this works 😅 are you sure we need both of these declarations though, wouldn't '127.0.0.1' have us covered?

nataliiazab replied:
I know 😅

I think '127.0.0.1' should be enough - I was overly cautious and included both of them 🤸‍♂️ . It is now removed :)

In my local supabase new items pictures are stored at the following path, which is just the same as mentioned in the document:
![image](https://github.com/user-attachments/assets/0d04a137-fad2-4d2a-aa6b-8d7400c5b6ab)
@camelPhonso Just to double check if it is the same for both of us - where are your images stored when you open items/ profile tables?

camelPhonso replied:
Just checked, this is the same for me yes 👍

camelPhonso reviewed (changes requested):
This is brilliant @nataliiazab - honestly great stuff and solving something that a few of us had missed entirely.

I've left one or two questions just for my own clarification really and some small suggestions to tweak readability and error handling.

The main thing is that if we need to actually create a bucket for this to work then we should either create a migration file (there's guidance on this in our documentation in the `BEFORE_YOUR_FIRST_ISSUE.md` file) or find out how to set this in the `config.toml` file - I'd look into the toml file first because it's a cleaner solution and quite likely that it can be done through there.

Lastly, I'd like to take the chance for us to add to our tests. Now that everything is set up to run in our CI nicely and we have reliable seed data we should take every opportunity to improve our coverage.

Maybe you could add to the `authorisation.cy.js` an `it()` block that checks editing the profile information, including uploading a picture and finding it on the screen?

nataliiazab commented on components/form/UploadImageInput.tsx:
```diff
@@ -11,7 +11,9 @@ type UploadImageProps = {
 const CDN =
-  'https://undfcbmldjkujposixvn.supabase.co/storage/v1/object/public/images/';
+  process.env.NODE_ENV === 'production'
+    ? 'https://undfcbmldjkujposixvn.supabase.co/storage/v1/object/public/images/'
+    : 'http://127.0.0.1:54321/storage/v1/object/public/images/';
```
Thanks, I love your suggestion to directly call `process.env`.  I was worried that someone would use a different local supabase URL that wouldn't work. But now we won't have that issue ⭐ ⭐ ⭐

camelPhonso commented on components/form/UploadImageInput.tsx:
```diff
@@ -54,19 +53,49 @@ const UploadImageInput: React.FC<UploadImageProps> = ({
 
-        if (error) {
-          console.log(error);
+          if (oldImages.length > 0) {
+            const { error: deleteError } = await supabase.storage
+              .from('images')
+              .remove(oldImages.map((image) => userId + '/' + image));
```
As I was going over this again to answer your question I spotted an issue that had gone overlooked in the first review - because you're now identifying the images with userId we are limiting all image uploads from that user to a single image - including for their items.

That means that if you upload a new profile picture you will replace the old one, which is what we wanted, but then if you upload a new item the image on that item will delete your profile picture as well

nataliiazab replied:
@camelPhonso - I've fixed it by adding 'profile_' prefix to the name of the profile image file. At first, I wanted to create separate buckets for the items-images and profile-images. However, I always encountered the same error -` UploadImageInput.tsx:73 {statusCode: '403', error: 'Unauthorized', message: 'new row violates row-level security policy'}` which I think is because of the differences in my local and prod database configurations.

Let me know what you think and if it fixes the problem :)

camelPhonso commented on components/form/UploadImageInput.tsx:
```diff
@@ -54,33 +56,38 @@ const UploadImageInput: React.FC<UploadImageProps> = ({
           }
 
-          const imageName = uuidv4();
+          const prefix = imageType === 'profile' ? 'profile_' : 'item_';
```
we could probably just interpolate the actual imageType since it's being controlled by the type declaration and it would make it a nicer read for the future

```typescript
const prefix = `${imageType}_`
```

nataliiazab replied:
nice suggestion - thank you!

nataliiazab replied:
... but after working on this for more than two hours, cannot make it work 😫. I added the following lines of code :
```javascript
console.log('imageType', imageType);
const prefix = `${imageType}_`;
```
But I keep getting "imageType undefined" for 'item'. I think there might be an issue with how the imageType prop is being handled from AddNewItemForm to UploadImageInput. I haven't been able to figure out what the problem is yet, but it's probably something very simple  😄 

I'll leave this as it is for now.

camelPhonso replied:
This is odd and makes me really curious, but I'm running everything fine and getting the expected urls for image uploads 🤔 

Since we're adding tests as well to check on this feature I feel pretty confident in what we have.

camelPhonso commented on components/form/UploadImageInput.tsx:
```diff
@@ -8,6 +8,7 @@ type UploadImageProps = {
   setImageSrc: (src: string) => void;
   setError?: (error: string) => void;
   isRequired?: boolean;
+  imageType: 'item' | 'profile';
```
Yep, this works and I think it's a pretty elegant solution. I'd say just add some information to the [JS Docs](https://github.com/enBloc-org/kindly/blob/dev/.github/BEST_PRACTICE.md#documenting-through-js-docs) for this component. At the moment it looks like it's actually missing details on some of the other props so if you could add those as well it would be good refactor of the documentation:

```typescript
/**
 * @description this function will add any image to our project storage.
 * @param isRequired is an optional parameter that accepts a boolean. It will consider the value of false unless explicitly assigned the value of true.
 */

const UploadImageInput: React.FC<UploadImageProps> = ({
  setImageSrc,
  setError,
  isRequired: isRequired = false,
  imageType,
}) => {
```

camelPhonso reviewed (changes requested):
It's looking great!

It all seems to work on my end, the only thing is that I can't find any way that we can force the local environment to create a bucket when we do `npx supabase start` or `npx supabase db reset`

That means that any time a developer restarts their container they will lose that bucket and run into an error if they try to upload an image.

I still think that having things set up so that we can actually test image uploads visually is a better option so what I think we should do is at least add to the documentation the steps to creating the bucket.

Can you edit `BEFORE_YOUR_FIRST_ISSUE.md` so that after setting up the containerised supabase developers go into the `STUDIO URL` and manually create a bucket called images and set it as `public` ?

It's worth also having a note to say that container will be lost if they reset the database or destroy the container.

nataliiazab commented:
Thank you @camelPhonso for your comments - learning a lot from you!

Just a quick update on my last commits:
- I've added a simple userProfile.cy.js test to check if a user can edit their profile username, profile images and if they are visible after reload
- Updated `BEFORE_YOUR_FIRST_ISSUE.md` on how to create an 'images' bucket for the local environment
- Added some information to JS docs comments for 'UploadImageInput.tsx' : `setImageSrc , setError , imageType`

camelPhonso commented on cypress/e2e/userProfile.cy.js:
```diff
@@ -0,0 +1,31 @@
+    });
+    cy.get('button').contains('EDIT PROFILE').click();
+    cy.reload();
+    cy.get('img[alt="User avatar"]')
+      .should('be.visible')
+      .and('have.attr', 'src');
```
Great! I think this part of the test can be improved a bit. If we check that the `src` value contains a part of the url that we are enforcing in the code we can verify that it's in fact displaying an image from our bucket.

I played around with it a bit and came up with the line below that should do it for us:

```js
cy.get('img[alt="User avatar"]')
      .should('be.visible')
      .invoke('attr', 'src')
      .then((src) => {
        expect(src).to.match(/public\/images/gi);
      });
```

camelPhonso commented on package.json:
```diff
@@ -59,6 +59,7 @@
     "@typescript-eslint/eslint-plugin": "^6.12.0",
     "@typescript-eslint/parser": "^6.12.0",
     "cypress": "^13.7.2",
+    "cypress-file-upload": "^5.0.8",
```
This seems fine to me but the `item.cy.js` tests have been uploading an image without this extension, is there a reason why we needed to add this?

nataliiazab replied:
Thanks for bringing this up, @camelPhonso! I initially installed this based on a YouTube tutorial I watched, but then I realised that this is for older versions of Cypress and implemented a code without it. 

I've removed the dependency with my new commit, as it is not being used by neither `userProfile.cy.js` nor ` item.cy.js`

camelPhonso reviewed (changes requested):
Awesome addition, thanks! My only note is an extra assertion in the new tests to make sure we are recovering our images from the bucket. Apart from that it looks great :)

nichgalzin commented on components/form/UploadImageInput.tsx:
```diff
@@ -8,20 +8,24 @@ type UploadImageProps = {
   isRequired: isRequired = false,
+  imageType,
 }) => {
   const supabase = createClientComponentClient();
```
Hey @nataliiazab , 
   Alphonso asked me to give your PR, a quick once over, and although you didn't right this bit I was wondering if you could refactor this quickly? Instead a creating a new supabase client like this we are always trying to use the function newClient() in the supabase/utils folder. 
   This is because this function first checks if there's an instance of this client so we don't create more then one as it was causing some error (this is called the singleton principle and a fun pattern to learn about). So if could update this for us, that would be amazing!

nataliiazab replied:
Hi @nichgalzin, thanks for reviewing my PR! 

❌ I keep receiving the following error when I change from createClientComponentClient() to newClient().
> ![image](https://github.com/user-attachments/assets/85095050-b966-4833-96ae-6083d3b63bc5)

I've been adjusting the bucket permission policies, but haven't made any progress. Could you please check if you're experiencing the same error? ☺️ 

Here is one of the changes I've tried:
> ![image](https://github.com/user-attachments/assets/e7af655e-8359-4561-9ebf-2a0947bbb211)

nichgalzin replied:
Hey @nataliiazab,
    Sorry for the slow reply on this. I'm having a play and not getting either version to work. I'll message you on discord, maybe we could hop on a call sometime this week and try to resolve things together?

nataliiazab replied:
@nichgalzin - I've added the changes we discussed. Thanks a million for your help and review ⭐

nichgalzin commented on components/form/UploadImageInput.tsx:
```diff
@@ -54,19 +58,54 @@ const UploadImageInput: React.FC<UploadImageProps> = ({
+          if (imageType === 'profile') {
+            const { data: listData, error: listError } = await supabase.storage
+              .from('images')
+              .list(userId, { recursive: true });
```
This line is causing a typescript error, although everything seems to be working. It seams like recursive is not a property we can use here, and it seems to work without it. So maybe this is not necessary. Besides these two changes everything looks great. Keep up the good work!

nataliiazab replied:
Thanks, @nichgalzin - this is redundant indeed. I've deleted now 🙌

nichgalzin reviewed (changes requested):
A couple little quick changes and then this code it good to go!

camelPhonso reviewed (approved):
Just saw all the adjustments Nich asked for are completed and this is working great!

Congratulations on closing 2 issues in one go - Approved ✨
